app/controllers/application_controller.rb

```diff
@@ -67,6 +67,24 @@ def get_background_id(sample)
 
   private
 
+  # Modified from https://gist.github.com/josevalim/fb706b1e933ef01e4fb6
+  def authenticate_user_from_token!
+    user_email = request.headers['X-User-Email'] || params[:user_token]
+    user_token = request.headers['X-User-Token'] || params[:user_email.presence]
+
+    user = user_email && User.find_by(email: user_email)
+
+    # Notice how we use Devise.secure_compare to compare the token
+    # in the database with the token given in the params, mitigating
+    # timing attacks.
+    if user && Devise.secure_compare(user.authentication_token, user_token)
+      sign_in user, store: false
+    # Redirect if no current user by other auth
```

Comment:
I'm not 100% sure if this is the existing behavior... please advise.

Reply:
I think it should raise an exception and cause the attempted CLI upload to fail (that's the only place where we use tokens for authentication) -- see https://github.com/chanzuckerberg/idseq-cli and https://idseq.net/cli_user_instructions for context

Reply:
According to the unit test, the existing behavior is to 30x redirect on failed token auth. My question was more around the possible interaction between normal session auth and token auth.

Reply:
I'm more used to use JWT bearer tokens, and I'm not familiar to `devise` gem yet, so I'm trying to understand how these CRSF tokens are being managed (looks like our authentication mechanism is also cookie based, right?)
I agree it would make sense returning a 401, although it looks like the current behavior is not doing that, at least for the web authentication, which is returning a 200 if the cookie is not present or is invalid. Maybe the CLI upload flow is different, but I couldn't find the code that is handling that.